Bump swoval version

eatkins opened the pull request:
I discovered a nasty performance bug that impacted windows and linux in
swoval. Whenever the background watch process for the global file
repository detected that a directory was modified, it would rescan the
entire directory to determine what had changed. This could cause a huge
amount of io to happen if there were a lot of files in the modified
directory. It also was generally unnecessary because the directory
should have been updated via incremental file updates. This change does
make it slightly more likely that the cache will get out of sync with
the file system, but the cache is easily flushed and users can disable
it if they have any issues anyway.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=develop&io=pull/222/head&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins commented:
I found another performance bug so I have to make yet another swoval version. Things are looking really good with my manual testing on windows though now.

eed3si9n commented:
Nice. Good to catch these before 1.3.x.

eatkins committed:
Bump swoval - long description of changes

This commit bumps swoval, but I think it's worth adding some additional
discussion about what's changed since 2.0.7.

1) The library now handles relative paths. In the sbt use case, this
   generally shouldn't matter since globs are automatically
   converted to absolute paths, but it's still useful to have this
   feature. While I do think that relative paths should generally be
   discouraged, there are valid uses cases for them.

2) The old implementation of FileTreeRepository had a performance bug
   where whenever it received a directory modification event, it would
   poll the directory (and all of its children) for updates. This made
   the library run very slowly on windows which has a lot of problems
   with concurrent io on the same files. I made it so that now a
   directory update event will on update the entry for that directory.
   We will rescan the entire directory tree if an overflow event is
   detected by the watch service, but these should be rare in the wild.

3) There were a few minor bugs that could lead to the cache getting into
   an incorrect state. These bugs were largely restricted to monitoring
   symlinks or absent files (the library lets you register a path that
   doesn't exist and it will walk up the file system tree until it finds
   a path that does exist and start monitoring that directory). It's
   worth pointing out though that both of these features do work in
   general. I've tested manually on mac/windows/linux and if you have,
   say, a symlink to a source file, the cache will detect if the
   target file changes and report an event for the _linked_ file.
   Watchman, for example, does not support this kind of symlink
   monitoring because it's basically impossible to get right for every
   use case. It is however, not that bad to implement for most common
   use cases. If a loop is detected, the cache stops following the link
   further but will not get caught in an infinite loop or crash.

   Symlink watching is really nice for cross platform projects because
   you can put the shared source files in a non-source directory and
   create links in the platform specific directories and still have
   watch work without polling.

4) I made the testing significantly more robust. It's worth describing
   the testing strategy for swoval in detail because if the swoval
   library doesn't work reliably, users will not like using the built in
   file system cache. The library is cross platform with a matrix that
   looks like:
     scala 2.12 -- windows/mac/linux (scala.js and java)
     scala 2.11 -- windows/mac/linux (scala.js and java)
     scala 2.10 -- windows/mac/linux java

   The scala.js sources are mostly generated from java sources using a
   fork of https://github.com/timowest/scalagen. I had to write platform
   specific code to manage concurrency since the jvm and js have
   completely different concurrency models. To support sharing the tests
   cross platform, I use the async test feature of utest. The way this
   works is that if a test returns a future, utest will wait until the
   future completes to finish the test. In an ideal world, the tests
   could be written so that there is a queue or latch that is modified
   in a callback and the test could then block on the queue or latch for
   some timeout. This is basically impossible in js though so instead I
   had to implement a queue and a latch whose blocking methods actually
   return futures. As a result, the tests are highly concurrent and also
   given the difference in concurrency models in js and on the jvm,
   callbacks tend to be evaluated in different orders in js vs jvm. Even
   though I don't think anyone is using the scala.js library, it is
   still valuable because it validates the caching logic under different
   concurrency conditions.

   In CI, I run the jvm tests with jacoco code coverage and have set
   failure limits for most of the parameters. FWIW, the line coverage is
   around 90% and the cyclomatic complexity is around 70%. Most of the
   missing lines are in exception handlers that are difficult to trigger
   but must be handled because the code calls apis that can but almost
   never do throw IOExceptions.

   I also was worried about flaky tests. In my experience, a flaky test
   always means there is a bug in either your test code or your main
   code. You really can't be sure until you track it down and remove the
   flakiness. To guard against flaky tests, after I run the normal test
   suite for js and the jvm. I run a second round of tests where I
   re-run the test suite some number of times. Right now the iteration
   count is set to 50 on mac and linux builds and 100 on windows builds
   (of which there are fewer in the matrix -- I only test on 2.12 for
   windows since appveyor is stingy with build agents). Moreover, when
   running each suite, I pick a random number of threads (that is biased
   towards the number of tests) and run groups of tests in parallel.
   This tries to ensure that the tests are run with enough concurrency
   to really shuffle the order of callback evaluation. If any of the
   iterations fail, the whole build fails. This is quite strict, but I
   think that it's worth it so that I can assume that the cache works
   most of the time.

5) Related to 4, I added debug logging to the swoval library. It can be
   enabled with the system properties swoval.debug and, to provide a
   custom logger, swoval.debug.logger. The debug logger will log all
   path registrations and unregistrations as well as every file event
   that is detected. This is very noisy, but useful if it seems like
   events are not being detected by the library. For example, I was able
   to determine that for one bug the low level file monitor was
   correctly detecting the file events, but they were not correctly
   forwarded to the cache. It is really hard to track these kinds of
   bugs down without logging.

eatkins committed:
Add private[sbt] to all file repository classes

I have rethought these interfaces and will almost certainly want to
change them. The easiest path forward is to set them all private[sbt]
so that I can change them in a future version without breaking binary
compatibility.
